Exception based error handling for parameter requirements.

PaulPyl opened the pull request:
I added some std::exception's to the code for the four assertions that check the mapping parameters. Maybe there are more assertions that could be replaced in a similar way? This has been the first C++ in a good number of years for me so maybe it needs some review on best practices etc.

That being said, I could compile it on my Intel MacBook Pro 16 inch (2019) model without trouble and I could get it to exit with the appropriate message when I used parameters outside the specified ranges.

I think the whole maximum seed length being map_param.max_dist + map_param.k is somewhat confusing. The requirement should probably be that max_dist is < 255 and the message should say that instead of adding k on both side of the equation.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
-    assert( ( map_param.s <= map_param.k ) && " s have to be smaller or equal to k!");
-    assert( ( (map_param.k-map_param.s) % 2 == 0) && " k - s have to be an even number to create canonical syncmers. Set s to e.g., k-2, k-4, k-6, k-8.");
-    assert(map_param.max_dist <= 255 && " -m (maximum seed length have to be smaller than 255 + k.");
+//    assert(map_param.k > 7 && "You should really not use too small strobe size!");
```
Just delete this. If we want to see how the code was before the change, we can always see it in the Git history.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
         << "E: " << opt.E << std::endl;
 
+    std::cerr << std::endl << "Verifying parameters:" << std::endl;
+    try
```
The current style in the code is that the `{` is on the same line as the `try`.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
+    {
+	map_param.verify();
+    }
+    catch(std::exception& e)
```
There should be a space after the `catch`

marcelm replied:
Ah, and I also just read http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#e15-throw-by-value-catch-exceptions-from-a-hierarchy-by-reference, which recommends to catch const references.

marcelm commented on src/exceptions.hpp:
```diff
@@ -0,0 +1,29 @@
+//
```
Would you mind leaving out this header? I know Kristoffer did it this way, however, it just contains redundant information. We can look up author and creation date (and much more) in the Git history. And we already know the name of the file because we opened it somehow (worse, if the file is renamed, we need to remember to update the comment).

marcelm commented on src/exceptions.hpp~:
```diff
@@ -0,0 +1,9 @@
+#include <exception>
```
This is also a backup file.

To prevent this, create a file `~/.config/git/ignore` and add the line `*~` to it (global, not project-specific ignore patterns). But we can also add a `.gitignore` file to the StrobeAlign repo.

marcelm commented on src/exceptions.cpp~:
```diff
@@ -0,0 +1,33 @@
+#include "exceptions.hpp"
```
You seem to have accidentally added a backup file (extension is `cpp~`)

marcelm commented on src/exceptions.hpp:
```diff
@@ -0,0 +1,29 @@
+
+#include <exception>
+
+class k_exception: public std::exception {
```
I think this could be much simpler. You can just throw a `std::runtime_error` with a message. Or introduce a subclass of `std::runtime_error`, but just one is enough.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
         << "O: " << opt.O << std::endl
         << "E: " << opt.E << std::endl;
 
+    std::cerr << std::endl << "Verifying parameters:" << std::endl;
```
This may just be here for debugging purposes, but I would suggest to remove this before merging. This just clutters the output which is already quite long (we need to clean this up at some point) and doesn’t add any information: If the parameters are fine, there’s no need to notify the user that they are, and if they aren’t, there’ll be an error message anyway.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
+    {
+	map_param.verify();
+    }
+    catch(std::exception& e)
```
This is too broad: If you catch `std::exception`, you catch every type of exception of the standard library. If on error occurred unrelated to the verification, you mask that error. Introducing a common base class for the verification exceptions would allow you to be more specific here.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
+    }
+    catch(std::exception& e)
+    {
+	std::cerr << "An exception occured in parameter verification: " << e.what() << std::endl;
+	std::cerr << std::endl << "Be advised of the following requirements:" << std::endl;
+	std::cerr << "- k should be from the interval [8,32], i.e. at least 8 and at most 32" << std::endl;
+	std::cerr << "- s should be smaller or equal to k" << std::endl;
+	std::cerr << "- (k - s) should be an even number" << std::endl;
+	std::cerr << "- the maximum seed length (defined as max_dist + k from parameters -m <max_dist> and -k <k>) should not exceed 255 + k" << std::endl;
```
The user should get a more specific error message, that is, you should print `e.what()`.

marcelm commented on src/main.cpp:
```diff
@@ -539,12 +541,27 @@ int main (int argc, char **argv)
+	std::cerr << "- s should be smaller or equal to k" << std::endl;
+	std::cerr << "- (k - s) should be an even number" << std::endl;
+	std::cerr << "- the maximum seed length (defined as max_dist + k from parameters -m <max_dist> and -k <k>) should not exceed 255 + k" << std::endl;
+	exit(1);
```
I’m not sure, but I think `exit()` may not clean up all resources correctly. Since this is within `main()`, you can just use `return EXIT_FAILURE`.

marcelm commented:
BTW, you can write 'Closes #44` in your commit message to automatically have GitHub close issue 44 when the PR with that commit is merged. I have now linked that issue manually (by clicking on "Development" on the right of this page).

PaulPyl commented:
I have addressed the comments in this commit, I chose to inherit from runtime_error and got rid of the backup files etc. Also I changed the name of the pull request since that was just the initial commit message before.

marcelm commented on src/main.cpp:
```diff
@@ -539,14 +541,14 @@ int main (int argc, char **argv)
+	map_param.verify();
+    }
+    catch (mapping_parameter_exception& e) {
+	std::cerr << "An exception occured in parameter verification: " << e.what() << std::endl;
```
This output is very programmer oriented. I think a better message would be sth. like: "A mapping parameter is invalid: ". A normal user doesn’t need to know that there was an exception, just what the problem is.

marcelm commented on src/exceptions.hpp:
```diff
@@ -0,0 +1,11 @@
+
+#include <stdexcept>
+
+class mapping_parameter_exception: public std::runtime_error {
```
Can you name this `BadMappingParameter` or so? See https://kevlinhenney.medium.com/exceptional-naming-6e3c8f5bffac for why I think dropping the `_exception` suffix is better.

Also, I’d like us to use CamelCase class names. I realize it’s not consistent with the rest of the code, and I haven’t brought it up, but if we do it, we might as well start here.

marcelm commented on src/main.cpp:
```diff
@@ -12,18 +12,20 @@
 #include <math.h>
 #include <inttypes.h>
 
+#include <stdexcept>
```
This is not needed anymore, correct?

marcelm replied:
Yes, but `main.cpp` doesn’t catch `runtime_error`, only `BadMappingParameter`. (At the moment, it doesn’t really make a difference because `main.cpp` includes `exceptions.hpp` which in turn includes `stdexcept`.)

PaulPyl replied:
I think I misunderstood that comment to mean to remove all dependencies on that o.O

marcelm commented on src/exceptions.hpp:
```diff
@@ -0,0 +1,11 @@
+#ifndef exceptions_hpp
```
Hm, now that this file is so small, I wonder whether it actually needs to be separate. You could just put the class into `index.hpp`. It’s an exception that is specific to the `mapping_params` struct anyway.

PaulPyl replied:
I don't have a string opinion on it either way, only that if there would be more exceptions added later (not related to mapping parameters), it might be nicer to have them all in one place instead of attached to whichever class they pertain to (or somewhere else if they are not class specific)?

marcelm replied:
Yeah, you’re right. I’m also undecided (both ways make sense), so just leave it where it is now.

PaulPyl commented:
Alright, I think this might be resolved at this point.

marcelm commented:
I agree, a nice step towards better error handling!
